stan/math/prim/mat/fun/gp_matern52_cov.hpp

```diff
@@ -0,0 +1,323 @@
+  T_l neg_root_5_inv_l = -1.0 * pow(5.0, 0.5) / length_scale;
+  typename return_type<T_x>::type sq_distance;
+  typename return_type<T_x>::type distance;
+
+  for (size_t i = 0; i < x_size; ++i) {
+    cov(i, i) = sigma_sq;
+    for (size_t j = i + 1; j < x_size; ++j) {
+      sq_distance = squared_distance(x[i], x[j]);
+      distance = sqrt(sq_distance);
+      cov(i, j) = sigma_sq
+                  * (1.0 + root_5_inv_l * distance + inv_l_sq_5_3 * sq_distance)
+                  * exp(neg_root_5_inv_l * distance);
+      cov(j, i) = cov(i, j);
+    }
+  }
+  return cov;
+}
+
+/**
+ * Returns a Matern 5/2 covariance matrix with one input vector
+ * with automatic relevance determination (ARD).
+ *
+ * \f[ k(x, x') = \sigma^2\bigg(1 +
+ *  \sqrt{5}\sqrt{\sum_{k=1}^{K}\frac{d(x, x')^2}{l_k^2}} +
+ *  \frac{5}{3} \sqrt{\sum_{k=1}^{K}\frac{d(x, x')^2}{l_k^2}}\bigg)
+ *  exp\bigg(-\frac{5}{3}\bigg(\sqrt{\sum_{k=1}^K{\frac{d(x, x')^2}{l_k^2}}
+ *  }\bigg)\bigg)
+ * \f]
+ *
+ * where \f$ d(x, x') \f$ is the Euclidean distance.
+ *
+ * @tparam T_x type of elements contained in vector x
+ * @tparam type of element of sigma, marginal standard deviation
+ * @tparam T_l type of elements in vector of length scale
+ *
+ * @param x std::vector of elements that can be used in stan::math::distance
+ * @param length_scale length scale
+ * @param sigma standard deviation that can be used in stan::math::square
+ * @throw std::domain error if sigma <= 0, l <= 0, or x is nan or inf
+ *
+ */
+template <typename T_x, typename T_s, typename T_l>
+inline typename Eigen::Matrix<typename return_type<T_x, T_s, T_l>::type,
+                              Eigen::Dynamic, Eigen::Dynamic>
+gp_matern52_cov(const std::vector<T_x> &x, const T_s &sigma,
+                const std::vector<T_l> &length_scale) {
+  using std::exp;
+  using std::pow;
+
+  size_t x_size = x.size();
+  size_t l_size = length_scale.size();
+  for (size_t n = 0; n < x_size; ++n)
+    check_not_nan("gp_matern52_cov", "x", x[n]);
+
+  check_positive_finite("gp_matern52_cov", "marginal variance", sigma);
+  check_positive_finite("gp_matern52_cov", "length scale", length_scale);
+  check_size_match("gp_matern52_cov", "x dimension", x[0].size(),
+                   "number of length scales", l_size);
+
+  Eigen::Matrix<typename return_type<T_x, T_s, T_l>::type, Eigen::Dynamic,
+                Eigen::Dynamic>
+      cov(x_size, x_size);
+
+  if (x_size == 0)
+    return cov;
+
+  T_s sigma_sq = square(sigma);
+  T_l root_5 = pow(5.0, 0.5);
+  T_l five_thirds = 5.0 / 3.0;
+  T_l neg_root_5 = -1.0 * pow(5.0, 0.5);
+  typename return_type<T_x>::type sq_distance;
+  typename return_type<T_x>::type root_sq_distance;
+
+  std::vector<Eigen::Matrix<T_l, -1, 1>> x_new(x_size);
+  for (size_t n = 0; n < x_size; ++n) {
+    for (size_t d = 0; d < l_size; ++d) {
+      x_new[n].resize(l_size, 1);
+      x_new[n][d] = divide(x[n][d], length_scale[d]);
+    }
+  }
+
+  for (size_t i = 0; i < x_size; ++i) {
```

Comment:
Since this loop is close to the one above, is there any reason this function isn't set up to construct the `x_new` and then call the other function? (It might need a little more work than that, but I think we can remove a lot of duplication.)

Reply:
@syclik good catch, yeah this is bad practice. I'm working on a functino in this PR at least in prim that allows to divide by the length scale, and it will be used in all the "special case of matern kernels"